### PR TITLE
add validation for ignore_attributes and default_target_attribute at …

### PR DESCRIPTION
**Pull request by a-moadel**
…craete_dataset

#### Reference Issue

#964

##### What does this PR implement/fix? Explain your changes.

1. First for ignore_attributes and default_target_attribute expand to list if they are comma sperated strings.
2. check for both attributes that all values are ie columns.

#### How should this PR be tested?

on call of create_dataset , we can try values for ignore_attributes and default_target_attribute that are part of the columns and some values that are not.

### CODE REVIEW

File: openml/datasets/functions.py [hunk 1]
```diff
@@ -333,6 +333,28 @@ def _load_features_from_file(features_file: str) -> Dict:
         return xml_dict["oml:data_features"]
 
 
+def _expand_parameter(parameter):
```
Comment (PGijsbers) [inline openml/datasets/functions.py:336]: please add type hints

File: openml/datasets/functions.py [hunk 2]
```diff
@@ -333,6 +333,28 @@ def _load_features_from_file(features_file: str) -> Dict:
...
+    expanded_parameter = []
+    if isinstance(parameter, str):
+        expanded_parameter = [x.strip() for x in parameter.split(",")]
+    elif isinstance(parameter, list):
+        expanded_parameter = parameter
+    return expanded_parameter
+
+
+def _validated_data_attributes(attributes, data_attributes, parameter_name):
```
Comment (PGijsbers) [inline openml/datasets/functions.py:345]: please add type hints

File: openml/datasets/functions.py [hunk 3]
```diff
@@ -333,6 +333,28 @@ def _load_features_from_file(features_file: str) -> Dict:
...
+    if attributes is not None:
```
Comment (PGijsbers) [inline openml/datasets/functions.py:346]: looking at the flow, it looks like `attributes` is never `None` since that is not a possible return value for `_expand_parameter`.

File: openml/datasets/functions.py [hunk 4]
```diff
@@ -333,6 +333,28 @@ def _load_features_from_file(features_file: str) -> Dict:
...
+        for attribute_ in attributes:
+            is_row_id_an_attribute = any([attr[0] == attribute_ for attr in data_attributes])
```
Comment (PGijsbers) [inline openml/datasets/functions.py:348]: The variable name does not make a lot of sense. It just indicates of the attribute is actually an attribute of the dataset, right?

File: tests/test_datasets/test_dataset_functions.py [hunk 5]
```diff
@@ -876,6 +876,94 @@ def test_get_online_dataset_format(self):
...
+                ignore_attribute=None,
+                citation=citation,
+                attributes="auto",
+                data=df,
+                version_label="test",
+                original_data_url=original_data_url,
+                paper_url=paper_url,
+            )
+
+    def test_create_dataset_ignore_attribute_validation(self):
```
Comment (PGijsbers) [inline tests/test_datasets/test_dataset_functions.py:923]: also a test case for row_id_attribute?

File: tests/test_datasets/test_dataset_functions.py [hunk 6]
```diff
@@ -876,6 +876,94 @@ def test_get_online_dataset_format(self):
             "The format of the ARFF files is different",
         )
 
+    def test_create_dataset_default_target_attribute_validation(self):
```
Comment (PGijsbers) [inline tests/test_datasets/test_dataset_functions.py:879]: I know we haven't used this functionality before in this project, but would you be able to refactor these similar tests using `pytest.mark.parametrize`? It allows you to easily change only a few variables with otherwise same test setup.

File: tests/test_datasets/test_dataset_functions.py [hunk 7]
```diff
@@ -909,7 +996,7 @@ def test_create_dataset_pandas(self):
             collection_date=collection_date,
             language=language,
             licence=licence,
-            default_target_attribute=default_target_attribute,
+            default_target_attribute="play",
```
Comment (PGijsbers) [inline tests/test_datasets/test_dataset_functions.py:999]: Why did you change these tests?

**Review by PGijsbers — CHANGES_REQUESTED** [review]
I left some questions and feedback that I'd like to see addressed 👍

File: tests/test_datasets/test_dataset_functions.py (hunk 7 above)
Comment (a-moadel) [reply tests/test_datasets/test_dataset_functions.py:999]: This one will fail because of the new logic , as "play" is not in the dataset columns.

File: openml/datasets/functions.py [hunk 8]
```diff
@@ -333,6 +333,27 @@ def _load_features_from_file(features_file: str) -> Dict:
         return xml_dict["oml:data_features"]
 
 
+def _expand_parameter(parameter: Union[str, list]):
```
Comment (PGijsbers) [inline openml/datasets/functions.py:336]:
```suggestion
def _expand_parameter(parameter: Union[str, List]) ->List[str]:
```

- `List` should be used as a type hint instead of `list` (`list` is only supported for type hints in Py3.9+)
- type hints include a return value (`-> None` of no return value)

File: openml/datasets/functions.py [hunk 9]
```diff
@@ -333,6 +333,27 @@ def _load_features_from_file(features_file: str) -> Dict:
...
+    expanded_parameter = []
+    if isinstance(parameter, str):
+        expanded_parameter = [x.strip() for x in parameter.split(",")]
+    elif isinstance(parameter, list):
+        expanded_parameter = parameter
+    return expanded_parameter
+
+
+def _validated_data_attributes(attributes: list, data_attributes: list, parameter_name: str):
```
Comment (PGijsbers) [inline openml/datasets/functions.py:345]: So here it should also say `-> None`. I am not adding it as a commit suggestion here, because I think it would violate the line length constraint we have. So best to do this manually and then have the `pre-commit` hook re-format if necessary.

File: tests/test_datasets/test_dataset_functions.py (hunk 7 above)
Comment (PGijsbers) [reply tests/test_datasets/test_dataset_functions.py:956]: But previously it was also set to `"play"`? And it says `column_names = ["rnd_str", "outlook", "temperature", "humidity", "windy", "play"]` just a few lines up. I am a bit confused about this. Maybe sitting together for this works better.

**Comment by PGijsbers** [comment]
I forgot about this earlier, could you also add a test which tests the new functionality that allows a number of attributes to be specified by a comma-separated string? (e.g. `ignore_attribute="windy,outlook"`

**Review by PGijsbers — APPROVED** [review]

**Comment by PGijsbers** [comment]
Thanks 🎉
